pylsp/plugins/jedi_completion.py

```diff
@@ -51,11 +51,19 @@
 # Types of parso node for errors
 _ERRORS = ('error_node', )
 
+# most recently retrieved completion items, used for resolution
+_LAST_COMPLETIONS = {}
+
 
 @hookimpl
 def pylsp_completions(config, document, position):
     """Get formatted completions for current code position"""
+    # pylint: disable=too-many-locals
+    # pylint: disable=global-statement
+    global _LAST_COMPLETIONS
```

Comment:
Is there another way to keep the last completion call instead of using a `global` variable here? We could create a centralized cache that hookspecs can use to store values and enters by parameter

Reply:
I can remove the global statement easily and use `.clear()` method instead easily but it would be marginally slower (garbage collector will be forced to clean up now rather than later which can push the task to idle time and give better response rate and completions are time-critical for UX). Would that be enough or do you actually want a persistent cache?

As for caching: yes, I originally though about it; there are three issues:
- we should not cache user-defined functions; the signature and documentation of those is subject to change and user will be confused and frustrated seeing the outdated suggestions; we have no reliable way of knowing whether a module they edit is their own or just something they import and forget about; we could be making educated guesses:
   - using the list of "big annoying libraries as in #26
   - trying to resolve the import path by scripting around the Python importlib machinery and make a guess based on the location (whether it includes sites-packages); I am not sure how reliable that would be
- storing inactive jedi completion items is likely to cause errors, certainly when using threads; I can try but my experience in other parts of the code was that it will just lead to a segfault equivalent
- it would require an invalidation mechanism to prevent memory leaks; I think that the best we can do is TTL, same as proposed for #26

Overall I am not convinced about the benefit of caching but we can try it. Do you think that this is worth trying it now on this PR or would it be easier to make small granular changes?

Reply:
> storing inactive jedi completion items is likely to cause errors, certainly when using threads; I can try but my experience in other parts of the code was that it will just lead to a segfault equivalent

To be clear we cannot store resolved completions because this defeats the purpose (performance gain from delaying the documentation/label resolution which provides a notable performance boost).

Reply:
Sorry for the confusion, I was referring wrongly to cache to a object that stores the last completion results that you can use on the following hookspec, I was thinking of this object as a reusable one that hookspecs can use to share information amongst them. The idea would be to have signatures like the following:

```python
@hookimpl
def pylsp_completions(config, document, position):
      information = document.shared_data  # or somename like that
      ...
      information['LAST_COMPLETIONS'] = ...

@hookimpl
def pylsp_completion_item_resolve(document, completion_item):
     last_completions = document.shared_data['LAST_COMPLETIONS']
     ...
```

This way we don't have global variables, while having a mechanism that other hookspecs can use to reuse information.

Reply:
Seems like a good idea! I will give it a try on the weekend.

Reply:
We cannot do that because there is no associated document for `completionItem/resolve`. While [`textDocument/completion`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion)s CompletionParams include text document identifier (as they inherit from [textDocumentPositionParams](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#textDocumentPositionParams)), the `completionItem/resolve` is context-less - it does not receive the document id so we cannot access the document.

Reply:
We could use the data entry to store the associated document:

```
/**
	 * A data entry field that is preserved on a completion item between
	 * a completion and a completion resolve request.
	 */
	data?: any;
```

Reply:
That's a good idea. There is a cost associated - we are transferring the document uris over the wire for each completion, but it should usually be shorter than e.g. a docstring, and always cheaper to compute than a docstring (not to mention other details). Plus there are two minor benefits:
- not leaking memory - once the document is close the last completion items should get garbage collected
- switching between documents there is a guarantee we will get correct resolution in case of two completions with the same label but different docstring/details.

Thank you for this suggestion and not giving up on it.

Reply:
Resolving as it now implemented in 79fc79fe004f98419ec09f0f07b7d61e391065a7.